### ISSUE TITLE
AOT support for tensorflow probability

### ISSUE THREAD

**Issue by volvador** [tensorflow/probability#481]
Does tensorflow probability has AOT compilation support?
For example, will I be able to compile a graph containing this op

    optim_results = tfp.optimizer.bfgs_minimize(fn, initial_position=start, tolerance=1e-8)

and have it produce executable code.

Indeed I have a use case where I need to do online optimizations (mainly input curve fitting)

What for the other ops (like probability distributions, pdfs, cdfs, quantiles etc..), do they have AOT support?

**Comment by brianwa84** [comment]
There is some support, but not consistent across the board. For example, I
am currently working on Gamma sampling support. I'd suggest trying this
pattern:

import tf.compat.v2 as tf
tf.enable_v2_behavior()

def f(args):
  return do_stuff(args)

@tf.function
def g(args):
  return tf.xla.experimental.compile(f, args)

g(args)

If you encounter issues with the compile, feel free to file on github & we
will try to fix as feasible.

Brian Patton | Software Engineer | bjp@google.com

**Comment by volvador** [comment]
Thanks. I am working with tf1.13 and `import tf.compat.v2 as tf` does not seem to work for me

**Comment by brianwa84** [comment]
TF 1.14 was recently released, and should have the symbols you'd need.

Brian Patton | Software Engineer | bjp@google.com

**Comment by brianwa84** [comment]
1.14 is compatible with TFP 0.7.0

Brian Patton | Software Engineer | bjp@google.com

**Comment by volvador** [comment]
Great. Thank you very much.
As I do not have 1.14 installed I went to Google colab and executed this code 
    
    import numpy as np
    import tensorflow.compat.v2 as tf
    tf.enable_v2_behavior()
    import tensorflow_probability as tfp
    def f():
        minimum = np.array([1.52, 1.12])  # The center of the quadratic bowl.
        scales = np.array([2.0, 3.0])  # The scales along the two axes.
        # The objective function and the gradient.
        def quadratic(x):
            value = tf.reduce_sum(scales * (x - minimum) ** 2)
            return value, tf.gradients(value, x)[0]

        start = tf.constant([0.6, 0.8])  # Starting point for the search.
        optim_results = tfp.optimizer.bfgs_minimize(
        quadratic, initial_position=start, tolerance=1e-8)
        return optim_results.position

    @tf.function
    def g():
         return tf.xla.experimental.compile(f)

    g()

And it gives the following answer:

    [<tf.Tensor: id=17018, shape=(2,), dtype=float32, numpy=array([1.52, 1.12], dtype=float32)>]

So bfgs seems to be supported.
Last question now : when do you think `xla.experimental` will be part of the official `xla`. 
And may you give me a manner to quickly test what is supported by the current `xla`, like the manner above?
Thanks a lot for help

**Comment by brianwa84** [comment]
I don't think xla.compile will be going away any time soon, but will ask
around abt API stability.

Can you clarify what you mean here?
to quickly test what is supported by the current xla, like the manner above?

Do you mean which parts of TFP?
Many distributions and bijectors should work, but not all. Mcmc should
work.

Intuitively, we should be able to get most anything to work that has
statically known tensor shapes.

**Comment by volvador** [comment]
Thank you.
What I meant is : I used `tf.xla.experimental.compile(f)` to check that all ops in the function f are supported (including the `tfp.optimizer.bfgs_minimize`).
As I used ``tf.xla.experimental.compile` the support is in `xla.experimental`, right?

Can I use the same trick to test what is supported by the `xla` (like coding a function `f` as above then try `tf.xla.compile(f)`

**Comment by brianwa84** [comment]
The main part of xla.compile that is experimental is that exact interface
to the compiler. The code that actually does the compilation is mostly
under tensorflow/compiler/{tf2xla,service} and is stable c++ code. On TF
GPU, auto-jit aromatically uses xla to fuse subsets of ops together (but
not control flow). Afaik, xla.compile is currently the only way to check
that *all ops* in f, including control flow, can be compiled.

So yes, it's a good way to probe whether the code in f is using only xla
compatible ops.

**Comment by brianwa84** [comment]
Going to close this, feel free to reopen for further discussion!
